packages/backend/src/core/FinalizeExitEventsCollector.ts

```diff
@@ -0,0 +1,122 @@
+import { decodeAssetId } from '@explorer/encoding'
+import { EthereumAddress, Hash256, StarkKey, Timestamp } from '@explorer/types'
+import { utils } from 'ethers'
+
+import { BlockRange } from '../model/BlockRange'
+import {
+  FinalizeExitAction,
+  ForcedTransactionsRepository,
+} from '../peripherals/database/ForcedTransactionsRepository'
+import { TransactionStatusRepository } from '../peripherals/database/TransactionStatusRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { getTransactionStatus } from './getForcedTransactionStatus'
+
+const PERPETUAL_ABI = new utils.Interface([
+  `event LogWithdrawalPerformed(
+    uint256 starkKey,
+    uint256 assetType,
+    uint256 nonQuantizedAmount,
+    uint256 quantizedAmount,
+    address recipient
+);`,
+])
+
+const LogWithdrawalPerformed = PERPETUAL_ABI.getEventTopic(
+  'LogWithdrawalPerformed'
+)
+
+type MinedTransaction = {
+  hash: Hash256
+  data: FinalizeExitAction
+  blockNumber: number
+  minedAt: Timestamp
+}
+
+export class FinalizeExitEventsCollector {
+  constructor(
+    private readonly ethereumClient: EthereumClient,
+    private readonly forcedTransactionsRepository: ForcedTransactionsRepository,
+    private readonly transactionStatusRepository: TransactionStatusRepository,
+    private readonly perpetualAddress: EthereumAddress,
+    readonly _getMinedTransactions?: (
+      blockRange: BlockRange
+    ) => Promise<MinedTransaction[]>
+  ) {
+    this.getMinedTransactions =
+      _getMinedTransactions || this.getMinedTransactions
+  }
+
+  async collect(
+    blockRange: BlockRange
+  ): Promise<{ added: number; updated: number; ignored: number }> {
+    const transactions = await this.getMinedTransactions(blockRange)
+    const results = await Promise.all(
+      transactions.map(async ({ hash, data, minedAt, blockNumber }) => {
```

Comment:
two nits to increase readability of this function:
1. insert some new lines
2. short circuit to avoid nesting, ie: `if(transaction) return 'ignored'`